parser.go

```diff
@@ -159,42 +164,46 @@ func parseRule(src, defaultValue string, rules *RuleList) string {
 	return defaultValue
 }
 
-func matchHistogramRules(labels *labelset, rules *HistogramRuleList) (matches []*labelset, matchOk bool) {
-	matches = make([]*labelset, 0)
+func parseHistograms(metricName string, srcLabels *labelset, rules *HistogramRuleList) (histos []*histogram, ok bool) {
+	histos = make([]*histogram, 0)
 
 	if rules == nil {
 		return
 	}
 
-	for _, r := range *rules {
-		if matchNames, ok := matchHistogramRule(labels, r); ok {
-			histLabels := &labelset{
-				Names:  make([]string, 0),
-				Values: make([]string, 0),
-			}
-
-			for _, name := range matchNames {
-				if val, ok := labels.Get(name); ok {
-					histLabels.Names = append(histLabels.Names, name)
-					histLabels.Values = append(histLabels.Values, val)
-				}
+	for _, rule := range *rules {
+		if metricName == rule.Metric {
+			if h, hOk := newHistogram(srcLabels, &rule); hOk {
+				histos = append(histos, h)
+				ok = true
 			}
-
-			matches = append(matches, histLabels)
-			matchOk = true
 		}
 	}
 
 	return
 }
 
-func matchHistogramRule(labels *labelset, rule HistogramRule) (matchNames []string, matchOk bool) {
-	matchNames = make([]string, 0)
+func newHistogram(srcLabels *labelset, rule *HistogramRule) (h *histogram, ok bool) {
+	h = &histogram{
+		Name: rule.Name,
+		Labels: &labelset{
+			Names:  make([]string, 0),
+			Values: make([]string, 0),
+		},
+	}
+
+	var sortedNames []string
+	for name, _ := range rule.Labels {
```

Comment:
~I assume you want `_, name` here. its `index, value`~

duh, map
then just do this to be more space efficient:
```
sortedNames := make([]string, 0, len(rule.Labels))
```

Reply:
actually maybe turn this into a function?
```
func keys(src map[string]string) []string {
  dst := make([]string, 0, len(src))
  for k, _ := range src {
    dst = append(dst, k)
  }
  sort.Strings(dst)
  return dst
}
```

then you can just do below:
```
for _, name := range keys(rule.Labels) {
```

Reply:
I like that